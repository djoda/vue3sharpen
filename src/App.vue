<template>
  <Sidebar v-if="logedIn" />
  <div class="main">
    <router-view></router-view>
  </div>
</template>

<script>
import { useStore } from "vuex";
import Sidebar from "@/components/Sidebar.vue";

export default {
  name: "App",
  components: {
    Sidebar,
  },
  setup() {
    let logedIn = false;
    if (localStorage.getItem("token")) {
      logedIn = true;
    } else logedIn = false;
    useStore().dispatch("getRates");

    return {
      logedIn,
    };
  },
  data() {
    return {};
  },
};
</script>

<style>
body {
  margin: 0;
  padding: 0;
  width: 100%;
  height: 100%;
}

#app {
  width: 100%;
  min-height: 100%;
}

.main {
  display: flex;
  justify-content: center;
  min-height: 100vh;
  width: calc(100vw - 250px - 15px);
  position: absolute;
  left: 260px;
  top: 0px;
  padding-left: 15px;
  background-color: rgb(250, 250, 250);
}
</style>
