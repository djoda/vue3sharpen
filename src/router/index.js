import { createRouter, createWebHistory } from 'vue-router'
import Products from "@/components/Products.vue"
import Discount from "@/components/Discount.vue"

const routes = [
  {
    path: "/",
    name: "Products",
    component: Products
  },
  {
    path: "/discount",
    name: "Discount",
    component: Discount
  }
]

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes
})

export default router
