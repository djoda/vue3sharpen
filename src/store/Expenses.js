class Expenses {
    constructor(description, amount, amonutType, productUPC) {
        this.Description = description;
        this.Amount = amount;
        this.AmountType = amonutType;
        this.ProductUPC = productUPC;
    }
} export default Expenses;