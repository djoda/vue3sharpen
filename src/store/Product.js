import { useStore } from "vuex";

class Product {
    constructor(name, UPC, price, image) {
        this.Name = name;
        this.UPC = UPC;
        this.Price = price;
        this.TaxRate = 20;
        this.UPCDiscount = 0;
        this.Image = image;
    }

    getUniversalDiscount() {
        return (useStore()).state.universalDiscount;
    }

    calculateGlobal() {
        return this.Price * this.getUniversalDiscount() / 100;
    }

    calculateDiscount() {
        if (useStore().state.additiveMode && this.UPCDiscount != 0) {
            return this.calculateGlobal() + this.calculateUCPDiscount();
        } else {
            return (this.Price - this.calculateGlobal()) * this.UPCDiscount / 100;
        }
    }

    calculateTax() {
        // console.log((useStore()).state.beforeDiscount)
        if ((useStore()).state.beforeDiscount)
            return this.Price * (this.TaxRate / 100);
        else {
            // console.log("ASDAS", this.calculateDiscount(this.getUniversalDiscount()))
            return (this.Price - this.calculateDiscount()) * this.TaxRate / 100;
        }
    }

    calculateUCPDiscount() {
        return this.Price * this.UPCDiscount / 100;
    }

    calculateExpenses() {
        let total = 0;
        for (let expense of useStore().state.expenses) {
            if (expense.AmountType === "%") {
                total += this.Price * expense.Amount / 100;
            } else {
                total += expense.Amount;
            }
        }
        return total;
    }

    calculateTotal() {
        return (
            (this.Price + this.calculateExpenses() - this.calculateDiscount() + this.calculateTax()).toFixed(2)
        );
    }
}
export default Product;