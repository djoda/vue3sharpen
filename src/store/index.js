import { createStore } from 'vuex'
import Product from "./Product"
import saveToLocalStorage from "./Plugins/webStorage"
import Expenses from './Expenses';

export default createStore({
  state: {
    mockProducts: [],
    products: [new Product("Product1", 111, 4565, "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"),
    new Product("Product2", 222, 3622, "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"),
    new Product("Product3", 333, 441565, "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg"),
    new Product("Product4", 444, 56, "https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg")],
    universalDiscount: 0,
    expenses: [new Expenses("Transport", 500, "$", 111)]
  },
  getters: {
    getByUPC: (state) => (UPC) => {
      return state.products.find(e => e.UPC === UPC);
    },
    getExpensesByUPC: (state) => (UPC) => {
      return state.expenses.filter(e => e.ProductUPC == UPC);
    }
  },
  mutations: {
    addProduct(state, item) {
      state.products.push(item);
    },
    setUniversalDiscount(state, ud) {
      state.universalDiscount = ud;
    },
    setMockProducts(state, items) {
      state.mockProducts = items;
    },
    setUPCDiscount(state, payload) {
      state.products = [...state.products.filter(e => e.UPC != payload.UPC), payload];
    },
    addExpense(state, payload) {
      state.expenses.push(payload);
      console.log(state.expenses)
    }
  },
  actions: {
    async fetchPosts({ commit }) {
      const products = await fetch("https://fakestoreapi.com/products").then(res => res.json());
      commit("setMockProducts", products);
    },
    async updateUPCDiscount(context, payload) {
      let prod = context.getters.getByUPC(payload.UPC);
      prod.UPCDiscount = payload.UPCDiscount;
      context.commit("setUPCDiscount", prod);
    },
    async updateProduct(context, payload) {
      context.commit("addProduct", payload);
    },
    async addExpense(context, payload) {
      context.commit("addExpense", payload);
    }
  },
  modules: {
  },
  plugins: [saveToLocalStorage]
})

