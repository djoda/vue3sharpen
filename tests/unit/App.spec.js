import App from "@/App.vue"
import { mount, } from "@vue/test-utils"
import store from "@/store"
import router from "@/router"

describe("App.vue", () => {
    test("snapshot", () => {
        const wrapper = mount(App, {
            global: {
                plugins: [store, router],
                mocks: {
                }
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    })
})