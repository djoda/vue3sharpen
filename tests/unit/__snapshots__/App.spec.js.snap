// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue snapshot 1`] = `
<div
  data-v-app=""
>
  
  <!--v-if-->
  <div
    class="main"
  >
    <!---->
  </div>
  
</div>
`;
